pkg/catalogv2/oci/client.go

```diff
@@ -0,0 +1,312 @@
+package oci
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	helmregistry "helm.sh/helm/v3/pkg/registry"
+	"helm.sh/helm/v3/pkg/repo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/helm/pkg/provenance"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/errcode"
+	"oras.land/oras-go/v2/registry/remote/retry"
+
+	catalogv1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+)
+
+// Client is a oci client which deals with helm charts for
+// fetching, storing that are inside OCI based helmregistries.
+type Client struct {
+	URL        string
+	repository string
+	registry   string
```

Comment:
What's the difference between registry and repository? Can you document these.

You normally have a Helm repository (the classic system with the index.yaml file) and an OCI registry. Within an oci registry you might have a namespace (see the spec [here](https://github.com/opencontainers/distribution-spec/blob/main/spec.md)).

Reply:
@mattfarina  The registry and repository here refer to the parts of the URL. 

In the oras go library a reference is defined as `registry`, `repository`, and `reference` (aka tag or digest) https://github.com/oras-project/oras-go/blob/main/registry/reference.go#L50

For understanding only
<img width="857" alt="Screenshot 2024-02-01 at 4 14 53 PM" src="https://github.com/rancher/rancher/assets/8402361/629dcf49-7aa9-4f99-b830-e078c2962815">

Image referenced from https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/

I have added docs. let me know if you have any questions.